### DISCUSSION TITLE
Deciding MapLibre scope

### DISCUSSION THREAD

**Discussion by nyurik** [maplibre/maplibre#27] [General governance]
This is a place for the MapLibre community to give feedback on the content of MapLibre umbrella organization, which features/repositories it should include, and how it should prioritize between projects.

We could be as conservative as "maplibre is for visualization only using maplibre styling language only, with no experimental stuff" to "maplibre should include anything that is map related and open source", or any position in between, e.g.  "focus on a simple to use full stack solution", "depend on how many people are willing to join and work on this feature", "donations should go towards X vs open source devs may contribute in other area", etc..  Based on this feedback the board should come up with the overall recommendations, which of course will be modified as the project evolves.

**Comment by wipfli** [comment]
I think MapLibre should help people making beautiful maps. The project should be about the software for achieving this goal. For me, this includes tools from OpenStreetMap data processing to rendering a map in the browser...

**Comment by msbarry** [comment]
I'd like to propose moving https://github.com/onthegomap/planetiler under MapLibre. I haven't had a chance to put a full proposal together but the idea is that MapLibre would provide a full-stack open source solution to visualize geographic data sources, all the way from raw data sources to an app or the browser.

**Comment by roblabs** [comment]
I have Swift & Javascript code for parsing GeoJSON in the [Indoor Mapping Data Format](https://register.apple.com/resources/imdf/) flavor.  IMDF is a structured, opinionated model which means GeoJSON `properties` are explicit and well known. 

---

Elevator Pitch:  _Good maps come from good data.  Reading good GeoJSON data makes good maps._

---

The proposed standalone repos would be software libraries that could be included in a project.

Side:  I use IMDF to model outdoor spaces like trailheads and wilderness.

**Comment by bdon** [comment]
I have a small project - [SDF Glyph Tool](https://github.com/protomaps/sdf-glyph-tool) - to generate the SDF glyph format from any font file. I can move this under the MapLibre umbrella, maybe under a new name like `maplibre-fonts`. The project doesn't have much relevance outside of MapLibre/Box GL projects. There are a few other alternative tools that mostly use the nodeJS bindings, but I think this is the only one that runs 100% in the browser.

**Reply by HarelM** [reply]
There's also this one that does the same on the server:
https://github.com/elastic/spritezero
Might be interesting to merge those projects and have a single javascript library that runs on both node and the browser that can manipulate things related to sdf fonts? This is just a thought...

**Reply by msbarry** [reply]
+1 - in the spirit of an open-source mapping stack you can depend on, there are a lot of little utilities like this you need to go from raw data to good looking maps. At the very least we might need to fork some unmaintained tools into maplibre, but the best case scenario is when we have a project under maplibre put together specifically to fill a need in the mapping stack.

**Reply by lseelenbinder** [reply]
We also have a tool for this: https://crates.io/crates/build_pbf_glyphs. Happy to port it over if folks are interested.

I think @bdon's is interesting since it's purely in the browser (JS).

**Reply by nyurik** [reply]
Sounds interesting, thx!  I suspect `-fonts` is a bit too generic, so maybe keep the name that describes the actual process like it was before, just in case we have alternative system?  That, or else we could have all font code together as subdirs in a single fonts monorepo?  Note though that fonts and icons might overlap in functionality, e.g. SDF icons are used in Elastic maps (maki icons, and thx @HarelM for the link), colored on the fly.  Lastly, there could be different implementations, e.g. nodejs vs rust ones, or the ones specific to webgl vs webgpu, etc.

**Reply by bdon** [reply]
Maybe we could do `maplibre-font-generator`; the implementation here uses the same code as `node-fontnik` but does not depend on the packaged Node binary, it uses `sdf-glyph-foundry` directly. 

I haven't checked the available fonts recently but I believe most of the existing pre-generated fonts are scattered around GitHub and do not include a reproducible way to build them from raw TTF files. So we could set up a gh-pages repo that includes packed ZIP downloads of popular fonts like Source/Noto Sans.

**Reply by maxammann** [reply]
Maybe we could also build the SDF font directly from Google Fonts in the browser :D

**Reply by nickpeihl** [reply]
> There's also this one that does the same on the server: https://github.com/elastic/spritezero Might be interesting to merge those projects and have a single javascript library that runs on both node and the browser that can manipulate things related to sdf fonts? This is just a thought...

FWIW, it is unlikely that we will continue to maintain this fork of spritezero. For Elastic Maps, we have moved from pre-rendering icons on the server to [rendering sdf icons on-the-fly in the browser](https://github.com/elastic/kibana/pull/119245). I'd be happy to see the community take ownership of spritezero.

**Comment by zstadler** [comment]
In terms of scope, I think the MapLibre umbrella should cover everything from taking OSM and other data to the many ways of providing maps.
New components need to be related to the existing components by consuming the outcome of other components, producing something for other components, or otherwise interact with them.

**Comment by maxammann** [comment]
My view: `MapLibre = FLOSS MapBox = FLOSS Vector and Raster Tile Rendering`

Which means anything related to vector and raster tile rendering should be part of MapLibre. Anything which is very low level should be part of [GeoRust](https://github.com/georust/) though. This would include geomentry types, GeoJSON parsing, projection stuff etc.

**Reply by nyurik** [reply]
Thx @maxammann, I agree that low level libs applicable to other projects may live in other places. georust is a bit too specific to Rust ecosystem, just like JTS is for Java. Neither of them would probably host something like planetiler though - as it is too specific to tile generation rather than a low-level geometry processing.  My biggest concern is that if we say "MapLibre == MapBox" we essentially let MapBox innovate, while also saying that no, we cannot introduce any fundamental improvements.  I don't think we should limit ourselves from large innovations (like rewriting everything in Rust and supporting alternative tile or font formats, etc).  I hope  https://github.com/maplibre/maplibre/discussions/27#discussioncomment-2749103 aligns with your thoughts (and I will be happy to correct it)

**Comment by klokan** [comment]
"MapLibre Open Maps SDKs" - that's how it is written on the website https://www.maplibre.org/
"Open-source mapping libraries for developers of web and mobile applications."

MapLibre word is often shown next to Leaflet, OpenLayers, Mapbox GL JS, next to other iOS & Android SDKs, sometimes Google Maps SDKs or Apple Maps SDKs - and this way people are not confused by what really is MapLibre - **MapLibre is a SDK**.
> SDK = "A software development toolkit (SDK) is a set of software tools and programs provided by hardware and software vendors that developers can use to build applications for specific platforms."

This is where we are now.

Let's keep the focus PLEASE - let's concentrate all the efforts and limited resources on doing one thing the best way in the world. 

Let's make great and stable rendering SDKs for developers - who will love and trust our visualizing components, for different platforms, with binding for different programming languages, producing output in different resolutions and formats - vector and raster - real time or for batch processing - components for geocoding, analytical processing, validations, experiments such as the great Rust experiment with WebGPU, experiments with PDF, bindings to the React/Svelte/Angular/Vue... frameworks, JS or Native, compatibility layers to other SDKs.

Most focus must go on JavaScript, Android and iOS - this is a must - but more components and more SDKs the better, as long as the 3 core are super stable and maintained really well.

This by itself is a HUGE goal...

MapLibre does not want to compete with its users - developers.
MapLibre should therefore deliver reusable open-source components - it should not deliver applications (beyond source code of demonstrators). We don't want end-user mobile app showing maps and competing with Bing or Google Maps for end-users. We don't need users to install MapLibre anyhow. We don't want MapLibre server application or kubernets application. MapLibre is an SDK - set of components and sample apps which developers use to create great applications.
This way - you are encouraging developers to start to work on their new mobile apps, on their new servers, on their new WHATEVER - while not competing with them, while staying neutral and while being open and supporting everybody - and while being present in multiple projects (which is beneficial for MapLibre SDKs - as there is more contributors to open-source components and less rivality, while components has a clear business model and motivation for contributors - developers - who are paid by users to develop applications). We should not randomly select one application and move it to organization - and give it a highlight of the MapLibre brand and community for some reason. This is why we have https://github.com/maplibre/awesome-maplibre ... people should be encouraged to develop their apps - and we are happy to link it - properly recomend and highlight with proud, but apps are not part of maplibre itself.

MapLibre does not want to compete with it's sponsors and donors, or community.
MapLibre is not OpenStreetMap / OpenMapTiles / OpenLayers / Apache / OSGEO etc.
MapLibre is not a free hosting provider for maps. MapLibre is not competing with sponsors and companies who use MapLibre SDKs for their business. We want them to feel encouraged to continue to recommend publicly MapLibre SDKs (without a need for wrappers) - and be encouraged to contribute in development time of their engineers or in money or other resources - and boost cooperation on the SDKs.
MapLibre is not OSGEO or foundation or umbrella for everything geo. MapLibre is not a "label" for your project telling "this is 100% open-source" - we don't have any validation processes for this in place. OSGEO does. They do it well.

Read the Memorandum - signed by the founders of this project - the goals remain to be the same: https://docs.google.com/document/d/1UB8VJ6_vMD9Zxfem7cunwa_G-hXo_ZRY1IeF5iGlh4w/edit

Now - we have to prove the world we as community can do the one thing - the SDKs - esp. JavaScript and Android and iOS really well - and deliver. This is why peple are joining - and why we are getting donations and sponsorships and why MapLibre are attractive for large number of companies and institutions and contractors and developers.

**Reply by nyurik** [reply]
Thanks @klokan, you do raise many important points, and surprisingly I believe we are mostly in agreement with what I wrote in https://github.com/maplibre/maplibre/discussions/27#discussioncomment-2749103. Just as you say, our primary focus should be SDKs for visualizing and interacting with maps. Yet, as I describe, if MapLibre is to survive and improve, we must be able to introduce fundamental changes to the ecosystem as needed. For example, extend MVT format for 3D buildings or support alternative formats, change font or icon formats for performance, extend tile metadata formats, etc.  This is not possible if we cannot modify all pieces of the stack.

I believe if we own SDKs for all relevant components, we will provide far greater value to all of our contributors. For example, we could provide an SDK to easily build a custom tile server. By default, it will be a very basic .mbtiles service, similar to `postserve` in OMT and `tileserver-gl` in MapTiler, but more stable, maintained, and able to support all advanced MapLibre-GL-JS capabilities. Most users will want to extend it to their needs, e.g. serve geojson, support geocoding, routing, or serve additional data from their own database, etc.  I do not believe this should conflict with our funding goals, while at the same time it definitely will attract more developers to our project, and will make sure MapLibre stays competitive in this space.

**Comment by nyurik** [comment]
_Thanks for all the amazing feedback!  Here is my **personal** perspective. I believe this discussion is critical to the future of MapLibre, so I tried to be thorough. I hope my thoughts align with the majority of the community. Replies are welcome :)_

---

I firmly believe that the community is far more important than code. Forking code is cheap. Building a community that wants to work together, innovate, respond to PRs and issues, help beginners, write documentation, report its progress, ... -- i.e. move the project forward is far harder.

Project scope could benefit us and harm us. If the scope is too strict, we won't be able to innovate in a significant way, causing stagnation and soon - irrelevance. Too broad, and the community loses focus, starts competing with itself and other FLOSS initiatives, thus wasting precious volunteer time.

**I believe focusing on the browser+native rendering/UI should be our priority. I also believe it is impossible to innovate without being able to improve the entire stack - from data to MVTs to serving to rendering to user interactions.**

We may want to extend the MVT standard to support 3D models, alternative projections, improved storage or parsing efficiency. We may have to use different font or tile binary formats to significantly improve performance. We may introduce variable zoom levels requiring client/server API improvements. We may want to define a full stack schema format so that projects like OpenMapTiles could describe how to convert OSM data to MVTs + styles in a standard way. We may want to support off-line (air-gapped) cases with incremental data updates. Scoping to just the tile rendering precludes us from such fundamental improvements.

### TLDR;
* Focus on rendering (gl-js/native) + plugins (directions/geocoding/...), but pursue the whole stack, e.g. tile generation, distribution, and serving. This also includes related data like defining tile metadata, font and icons tools, etc.
* Focus on SDKs, so that users can build their own web site, app, custom tile generation and tile serving infrastructure, etc.
* Avoid "not invented here" syndrome. If MapLibre uses a well maintained FLOSS component, try to work with maintainers for any needed changes instead of forking it. Only fork if the external project is dead or if we agree to go in a different direction from upstream.
* Avoid the actual "data". MapLibre is not a tile (data) service. It's a set of tools and libraries for any web site or app that needs a map.

**Comment by ZeLonewolf** [comment]
I want to start by echoing the excellent comments by @nyurik, @klokan, and @wipfli.  As a hobbyist myself, I'm awed by the amazing innovations this open source community produces from both individual volunteers and companies working together.  I first got involved in OpenStreetMap because I saw the intense value in an open data set and open tools because I recognized that the applications and services that those two things power aren't just trivialities.  They're innovations that _make the lives of ordinary people better_.  Individuals can't do those things alone because free time is limited. Companies can't do it alone because developers cost money which is most effectively directed towards creating the kinds of solutions that are not achievable by individuals in their spare time.  We're more impactful together and I think clearly defining the mapLibre scope as an emerging organization is a great way to set the stage for the future.

The scope of MapLibre ought to be about **tools**, **libraries**, **SDKs**, and **APIs**, and should be non-opinionated with respect for how users stack together those building blocks to create the "beautiful maps" envisioned by  @wipfli.  I would consider the entire process of going from data (both open source and bespoke) to digital maps in the user's hands to be fair game, and mapLibre should feel empowered to include things beyond front-end rendering if, when, and where it makes sense to do so if it unlocks barriers to better produce beautiful maps.

I agree with @nyurik that we should not reinvent the wheel here.  Stand on the shoulders of others, avoid competitive forking with other open source efforts, and think strategically about where the barriers to beautiful maps are and work to break them down.  We should ensure as a value that we're using open standards everywhere.  This allows both hobbyists and companies to most easily repurpose what we build.

**What mapLibre isn't**

I agree with @klokan that mapLibre should not be in competition with its users.  Producing and hosting maps, data set, fonts, and icons are powerful enablers for mapmaking but clearly out of scope -- that creative work squarely belongs to the users of mapLibre.  Companies such as MapTiler and Mapbox have led the way in doing these things, and I'm thrilled when our little open source hobby powers serious, sustainable, and production-quality products.  The community map project that I started, [OpenStreetMap Americana](https://github.com/ZeLonewolf/openstreetmap-americana), would not have been possible without the key innovations that these two companies have brought to the table; their continuing success is also our continuing success.  Open source is used by companies everywhere, whether or not they know it.  Making open source better makes the world better, and that's not just through hobbyshop projects!

Lastly, mapLibre should not be in the business of developing data schemas which determine how individual elements of map data become presented.  Schema development belongs in the hands of people making maps, not people making map tools.  Projects like OpenMapTiles have pioneered the power of common data schemas for building maps.  These are good things, because they provide a launch point for map makers as well as, in the case of OpenStreetMap, providing mapper feedback to the many volunteers putting data in the map in the first place.

**Comment by wipfli** [comment]
Ah I really like this community. I'm proud that we can have such a good discussion. Thanks everyone!

**Comment by lseelenbinder** [comment]
Really good points made by all. I echo @wipfli and @nyurik, and say I really appreciate everyone who's taking the time to add their thoughts!

My initial vision for MapLibre was much like @klokan's: a solid set of SDKs on which developers could build really amazing maps. I agree we don't need to (and shouldn't) build consumer-facing applications.

I would like to voice strong support for @wipfli's, @nyurik's and @ZeLonewolf's ideas that the scope should be about tooling, libraries, and APIs. I think MapLibre could (and ideally should) become an excellent one-stop shop for OSM-based maps (just as OpenStreetMap is a one-stop shop for open map data). This, in my opinion, involves everything from tools to develop and build schemas (e.g., planetiler) to client-side rendering SDKs (e.g., maplibre-rs).

I believe the focus should be on a vibrant community, so when projects ask for inclusion, and we gauge there is a large community interest in them, they should be included, if they fit within the tools/SDKs/libraries umbrella. In my opinion, a strong community is more important than a strict limit on what can or should be included in MapLibre, so if something _almost_ fits and has strong support, we should include it.

Ranking priorities:
1. A healthy, robust, and well-supported community: we can debate civilly, build together, and make amazing maps.
1. Vendor-agnostic SDKs for web, mobile, and server/desktop platforms.
1. Strong tooling to enable developers to build amazing maps with those SDKs.
1. Experimental projects (with ample community involvement) to push open-source mapping forward outside the interests of one company or another.
1. End-to-end maps from OSM data (think: schema -> building -> rendering, but very likely not building directly or hosting).

In the end, if MapLibre can host all that is necessary to build a map using OSM, that would be a first (to my knowledge) in open source under one organization, and that would be an incredible accomplishment, of which we could all be proud.

**Comment by nyurik** [comment]
After several multi-hour sessions, the MapLibre board came up with this scoping guideline. Your feedback is needed. Thanks!

> MapLibre develops open source mapping libraries and tools for others to build applications and services.
>
> Our primary goal is to build a friendly and innovative open-source community around the vendor-neutral MapLibre project.
>
> MapLibre aims to develop code. We prefer to empower our users to focus on designing maps, running map hosting services, distributing data and building end-user applications, rather than making such things part of MapLibre itself.
>
> The “MapLibre” brand name should be connected with the rendering libraries.

**Reply by wipfli** [reply]
If someone should be wondering who is on the MapLibre Governing Board: It is @nyurik @klokan @lseelenbinder and @wipfli at the moment, and there will be a community election for the board members this autumn.

**Reply by maxammann** [reply]
I really like the guideline. The only rhing which came to my mind is whether we should define "vendor-neutral". It would mean different things like:

* dont benefit a single maps vendor
* avoid funding from map vendors
* etc.

In that case I would also say that it is also unclear what a vendor is.

**Reply by nyurik** [reply]
Thx @maxma I think "user" is a better term here -- the goal is to produce code that would benefit everyone, without working too much in the interest of a single user (or the opposite -- NOT doing something in the interest of a single user while causing pain and suffering to all other users).

**Reply by msbarry** [reply]
I like it, sounds like to summarize:

- in scope: open source mapping libraries and tools
- out of scope: maps, applications, and services (maplibre makes it easier for others to do this)

I agree "vendor" is kind funny here - the goal is to be controlled by/benefitting the community of people/companies and not a single person/company. The corporate entities involved now just happen to be data vendors.

Also what is the reasoning behind the last line?

> The “MapLibre” brand name should be connected with the rendering libraries.

**Reply by maxammann** [reply]
> The “MapLibre” brand name should be connected with the rendering libraries.

Maybe we should say here: "MapLibre" brand name should be used in all MapLibre owned projects.

> open-source

I would favor to use "free and open-source" here. Or maybe use the term "FLOSS": https://opensource.stackexchange.com/questions/262/what-is-the-difference-between-foss-and-floss
Mapbox is open-source bot not FLOSS software in my understanding.

**Reply by wipfli** [reply]
Thanks for the inputs. We think reserving the brand name "MapLibre" for the rendering libraries:
* helps users to connect MapLibre with the rendering projects in MapLibre, at the moment this means predominantly MapLibre GL JS and Native
* emphasizes that the rendering libraries are at the core of MapLibre.

This is really about how we use the word MapLibre in the project names. For example, we think that the following are good usage examples: maplibre/maplibre-gl-js, maplibre/maplibre-gl-directions. (rendering library, plugin for rendering library). 

Example where the word maplibre should not be in the project name: maplibre/tileserver (the tile server is not directly connected to the rendering libraries). Note that the tileserver is still welcome to live under the MapLibre umbrella.

**Reply by ZeLonewolf** [reply]
It would be helpful to define what functionality is in scope for "rendering".  With respect to vector technology, is the intent to explicitly exclude utilities involved in mvt generation?

**Reply by wipfli** [reply]
@ZeLonewolf yes, this is correct regarding the GitHub repo name. Say that Planetiler would hypothetically would like to be under the MapLibre umbrella, then it would probably get the name `maplibre/planetiler` and not the name `maplibre/maplibre-tile-generator`. 

Note that with the above scope statement, only the map-style-independent part of Planetiler should be part of MapLibre.

**Reply by ZeLonewolf** [reply]
Thanks @wipfli that does make sense for maplibre scope, but I do find the terminology confusing.  If planetiler is not rendering vector tiles, what do we call that side of the technology stack?

**Reply by ZeLonewolf** [reply]
If it could be clarified as "client-side rendering" to distinguish it from server-side rendering, I think that would be a lot clearer.

**Reply by wipfli** [reply]
Regarding vendor, thanks for pointing out that this is maybe not the best term. As @nyurik suggested, we could maybe replace it with the word "user" and use the definition from http://oss-watch.ac.uk/resources/meritocraticgovernancemodel for the user:

>Users
Users are community members who have a need for the project. They are the most important members of the community and without them the project would have no purpose. Anyone can be a user; there are no special requirements.

>The project asks its users to participate in the project and community as much as possible. User contributions enable the project team to ensure that they are satisfying the needs of those users. Common user contributions include (but are not limited to):

>evangelising about the project (e.g. a link on a website and word-of-mouth awareness raising)
informing developers of strengths and weaknesses from a new user perspective
providing moral support (a ‘thank you’ goes a long way)
providing financial support (the software is open source, but its developers need to eat)

Marcel Normann was drafting the MapLibre Community document https://github.com/maplibre/maplibre/blob/main/COMMUNITY.md based on the meritocratic governance model and I think it can be a good source of inspiration for us. Personally, I like the sentence "a thank you goes a long way".

**Reply by wipfli** [reply]
Yes, I agree that the term rendering can also be used for tile generation. But if we say "client-side-rendering", don't we exclude rasterization on a server with MapLibre GL Native or so?

**Reply by wipfli** [reply]
I was hoping that we could use the term "pixel generation" but maybe this is not ideal...

**Reply by nyurik** [reply]
In my opinion, rendering means converting data to pixels. It can happen server side too. Planetiler transforms data, but doesn't generate visuals. Maplibre native or rs create pixels, and can run client and server side. 

The only trickier one in this case is some system that only works with imagery, like satellite pics, but let's discuss the name of such system if we ever get there ;)

Basic idea - if someone says "I use maplibre on my site", it should be fairly clear that they mean gl-js, not that they generated data with planetiler but used openlayers to show it. At the same time the maplibre organization clearly has wider scope and dev appeal, and should include the whole vertical stack of libs and tools to help build such site.

**Reply by ZeLonewolf** [reply]
Thanks @wipfli and @nyurik.  This was very helpful to understand how you're mentally labelling the parts of the vector tile technology stack, and how that is framing the scope of maplibre.  In line with that thinking, I've posted onthegomap/planetiler#238, which updates the planetiler documentation to be consistent with that vision.  While "render" has definitely been used in a much looser context previously, it appears from this discussion that we should probably be more deliberate with that terminology so everyone's on the same page.
